File: CourseRegistrationSystem/src/Domain/Lecturer.java
package Domain;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
@JsonIgnoreProperties(ignoreUnknown = true)
public class Lecturer extends Staff {

    private Collection<String> courseIds;

    private Collection<Course> courses;

    public Lecturer(String FName, String LName, Date birthdate, String staffNo) {
        super(FName, LName, birthdate, staffNo);
        setCourses(courseIds);
    }

    public Lecturer(){
        super();
    }


    public Collection<String> getCourseIds() {
        return courseIds;
    }

    public void setCourses(Collection<String> courseIds) {
        this.courseIds = courseIds;
    }

    public void addTeachingCourse(String courseId){
        courseIds.add(courseId);
    }

    @JsonIgnore
    public ArrayList<Course> getCourses(){
        ArrayList<Course> courses = new ArrayList<>();
        ArrayList<Course> allCourses = (ArrayList<Course>) Department.getInstance().getAllCourses();

        for(Course course : allCourses){
            if(this.getCourseIds().contains(course.getCourseCode())){
                courses.add(course);
            }
        }
        return courses;
    }

    public void setCourses(ArrayList<Course> courses){
        this.courses = courses;
    }

    @Override
    public String toString() {
        return "Lecturer{} " + super.toString();
    }
}

File: Iteration3/Domain/Lecturer.py
from typing import Collection

from Domain.Staff import Staff

class Lecturer(Staff):
    def __init__(self, first_name, last_name, staff_no, course_ids,password):
        super().__init__(first_name, last_name, staff_no, password)
        self.course_ids = course_ids
        from Domain.Department import Department


    def search(self, query):
        # Placeholder implementation for the abstract search method
        # Add your search logic here
        pass

    def get_course_ids(self):
        return self.course_ids

    def set_courses(self, course_ids: Collection[str]):
        self.course_ids = course_ids

    def add_teaching_course(self, course_id):
        self.course_ids.append(course_id)

    def get_courses(self):
        from Domain.Department import Department
        courses = []
        all_courses = Department.get_instance().get_all_courses()

        for course in all_courses:
            if course.get_course_code() in self.get_course_ids():
                courses.append(course)

        return courses

    def __str__(self):
        return f"Lecturer{{}} {super().__str__()}"


import json

class LecturerEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Lecturer):
            return {
                'first_name': obj.get_first_name(),
                'last_name': obj.get_last_name(),
                'staff_no': obj.get_staff_no(),
                'course_ids': obj.get_course_ids()
            }
        return super().default(obj)
